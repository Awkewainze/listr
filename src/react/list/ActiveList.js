import React from 'react';
import './ActiveList.scss';
import ListItem from './item/ListItem';
import TransitionGroup from 'react-transition-group/cjs/TransitionGroup';
import AppCSSTransition from '../common/AppCSSTransition';

const ActiveList = ({ items, onDelete, onToggle }) => (
  <TransitionGroup className="ActiveList">
    {items.map((item, i) =>
      <AppCSSTransition key={item.id} prefix="ListItem" timeout={200}>
        <ListItem
          key={item.id}
          item={item.text}
          checked={item.checked}
          onDelete={onDelete}
          index={i}
          itemId={item.id}
          onCheck={onToggle}
        />
      </AppCSSTransition>
    )}
  </TransitionGroup>
);

export default ActiveList;