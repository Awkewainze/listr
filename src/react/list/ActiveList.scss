@import '../../variables';
@import '../../mixins';

.ActiveList {
  .ListItem {
    transition: opacity 200ms ease;

    @include transitionState($exiting) {
      opacity: 0;
    }
  }
}