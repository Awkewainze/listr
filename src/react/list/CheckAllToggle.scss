@import "../../mixins";

.CheckAllToggle {
  @include gridLayout();
  grid-column: 1 / 5;
  align-items: center;
  margin: 12px 0;

  .Checkbox {
    grid-column: 2 / 3;
  }

  .CheckAllToggle__textWrap {
    grid-column: 3 / 4;

    .CheckAllToggle__textLabel {
      font-style: italic;
      font-size: 14px;
      cursor: pointer;
      user-select: none;
    }
  }
}