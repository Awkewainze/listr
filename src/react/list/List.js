import React from 'react';
import './List.scss';
import TransitionGroup from 'react-transition-group/cjs/TransitionGroup';

/*
 * This component will throw a warning during deletion of items.
 * See https://github.com/reactjs/react-transition-group/issues/429
 * and https://github.com/reactjs/react-transition-group/issues/606
 * We need to wait for react-transition-group to come to a solution to resolve this.
 */
const List = ({ children }) => (
  <TransitionGroup className="ActiveList">
    {children}
  </TransitionGroup>
);

export default List;