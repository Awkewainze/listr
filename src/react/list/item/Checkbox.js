import React from 'react';
import './Checkbox.scss';

const Checkbox = ({
  className,
  checked,
  index,
  onChange
}) => (
  <div className={`Checkbox ${className}`}>
    <input className="Checkbox__toggle" type="checkbox" id={`Checkbox-${index + 1}`} onChange={onChange} checked={checked}/>
    <label className="Checkbox__toggleLabel" htmlFor={`Checkbox-${index + 1}`}/>
  </div>
);

export default Checkbox;