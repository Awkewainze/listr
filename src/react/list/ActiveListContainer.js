import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import selectActiveList from '../../redux/selectors/list/selectActiveList';
import selectActiveListId from '../../redux/selectors/list/selectActiveListId';
import ContentWrap from '../common/ContentWrap';
import NoListFallback from './NoListFallback';
import { toggleItem, deleteItem } from '../../redux/actions/list';
import NewListItemPromptContainer from './create/NewListItemPromptContainer';
import ListItem from './item/ListItem';

const ActiveListContainer = () => {
  const list = useSelector(selectActiveList);
  const activeList = useSelector(selectActiveListId);
  const dispatch = useDispatch();

  if (list == null) {
    return (
      <ContentWrap>
        <NoListFallback />
      </ContentWrap>
    );
  }

  const onToggle = itemId => {
    dispatch(toggleItem(activeList, itemId));
  };

  const onDelete = itemId => {
    dispatch(deleteItem(activeList, itemId));
  };

  return (
    <ContentWrap>
      <div className="ActiveListContainer">
        <div className="ActiveListContainer__list">
          { list.map((item, i) => <ListItem key={i} item={item.text} checked={item.checked} onDelete={onDelete} index={i} itemId={item.id} onCheck={onToggle} />) }
        </div>
        <NewListItemPromptContainer/>
      </div>
    </ContentWrap>
  );
};

export default ActiveListContainer;