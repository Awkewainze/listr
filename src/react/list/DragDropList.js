import React from 'react';
import { useDispatch } from 'react-redux';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';
import List from './List';
import AppCSSTransition from '../common/AppCSSTransition';
import ListItem from './item/ListItem';
import { reorderItem } from '../../redux/actions/list';

const DragDropList = ({
  items,
  listId,
  onDelete,
  onCheck
}) => {
  const dispatch = useDispatch();

  const onDragEnd = result => {
    const { destination, source, draggableId } = result;

    if (!destination) {
      return;
    }

    if (destination.index === source.index) {
      return;
    }

    dispatch(reorderItem(listId, draggableId, source, destination));
  };

  return (
    <DragDropContext
      onDragEnd={onDragEnd}
    >
      <Droppable droppableId={listId}>
        {provided => (
          <List {...provided.droppableProps} innerRef={provided.innerRef}>
            {items.map((item, i) =>
              <AppCSSTransition key={item.id} prefix="ListItem" timeout={200}>
                <Draggable draggableId={item.id} index={i}>
                  {provided => (
                    <ListItem
                      key={item.id}
                      item={item.text}
                      checked={item.checked}
                      onDelete={onDelete}
                      index={i}
                      itemId={item.id}
                      onCheck={onCheck}
                      draggableProps={provided.draggableProps}
                      dragHandleProps={provided.dragHandleProps}
                      innerRef={provided.innerRef}
                    />
                  )}
                </Draggable>
              </AppCSSTransition>
            )}
            {provided.placeholder}
          </List>
        )}
      </Droppable>
    </DragDropContext>
  );
};

export default DragDropList;