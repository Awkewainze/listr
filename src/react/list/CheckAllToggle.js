import React from 'react';
import './CheckAllToggle.scss';
import Checkbox from '../common/Checkbox';
import { useSelector, useDispatch } from 'react-redux';
import selectActiveListAllChecked from '../../redux/selectors/list/selectActiveListAllChecked';
import selectActiveListId from '../../redux/selectors/list/selectActiveListId';
import { toggleAllChecked } from '../../redux/actions/list';

const CheckAllToggle = ({ id }) => {
  const activeList = useSelector(selectActiveListId);
  const allChecked = useSelector(selectActiveListAllChecked);
  const dispatch = useDispatch();

  const onChange = () => {
    dispatch(toggleAllChecked(activeList));
  };

  return (
    <div className="CheckAllToggle">
      <Checkbox
        checked={allChecked}
        id={id}
        onChange={onChange}
      />
    </div>
  );
};

export default CheckAllToggle;