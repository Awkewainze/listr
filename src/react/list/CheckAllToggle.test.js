import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import selectActiveListAllChecked from '../../redux/selectors/list/selectActiveListAllChecked';
import selectActiveListId from '../../redux/selectors/list/selectActiveListId';
import { shallow } from 'enzyme';
import CheckAllToggle from './CheckAllToggle';
import Checkbox from '../common/Checkbox';
import { toggleAllChecked } from '../../redux/actions/list';

jest.mock('react-redux', () => {
  return {
    __esModule: true,
    useSelector: jest.fn(),
    useDispatch: jest.fn()
  };
});
jest.mock('../../redux/selectors/list/selectActiveListAllChecked');
jest.mock('../../redux/selectors/list/selectActiveListId');
const mockDispatch = jest.fn();

const TEST_PROPS = {
  id: 'foo'
};

const expectedListId = 'list1234';

beforeEach(() => {
  useSelector.mockReset();
  useDispatch.mockReset();
  mockDispatch.mockReset();
  selectActiveListAllChecked.mockReset();
  selectActiveListId.mockReset();
  selectActiveListId.mockReturnValue(expectedListId);

  useSelector.mockImplementation(selector => selector());
  useDispatch.mockReturnValue(mockDispatch);
});

it('matches snapshot', () => {
  selectActiveListAllChecked.mockReturnValueOnce(true);
  expect(shallow(<CheckAllToggle {...TEST_PROPS} />)).toMatchSnapshot();
});

it('is unchecked if allChecked returns false', () => {
  selectActiveListAllChecked.mockReturnValueOnce(false);
  const wrapper = shallow(<CheckAllToggle {...TEST_PROPS} />);
  expect(wrapper.find(Checkbox).prop('checked')).toBeFalsy();
});

it('is checked if allChecked returns true', () => {
  selectActiveListAllChecked.mockReturnValueOnce(true);
  const wrapper = shallow(<CheckAllToggle {...TEST_PROPS} />);
  expect(wrapper.find(Checkbox).prop('checked')).toBeTruthy();
});

it('dispatches correct values', () => {
  selectActiveListAllChecked.mockReturnValueOnce(true);
  const wrapper = shallow(<CheckAllToggle {...TEST_PROPS} />);
  wrapper.find(Checkbox).prop('onChange')();
  expect(mockDispatch).toHaveBeenCalledTimes(1);
  expect(mockDispatch).toHaveBeenCalledWith(toggleAllChecked(expectedListId));
});

it('dispatches if label is clicked', () => {
  selectActiveListAllChecked.mockReturnValueOnce(true);
  const wrapper = shallow(<CheckAllToggle {...TEST_PROPS} />);
  wrapper.find('label.CheckAllToggle__textLabel').simulate('click');
  expect(mockDispatch).toHaveBeenCalledTimes(1);
  expect(mockDispatch).toHaveBeenCalledWith(toggleAllChecked(expectedListId));
});

