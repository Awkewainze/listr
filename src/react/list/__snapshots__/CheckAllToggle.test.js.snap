// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="CheckAllToggle"
>
  <Checkbox
    checked={true}
    id="foo"
    onChange={[Function]}
  />
  <div
    className="CheckAllToggle__textWrap"
  >
    <div
      className="CheckAllToggle__textLabel"
      onClick={[Function]}
    >
      Check / uncheck all
    </div>
  </div>
</div>
`;
